Key store v2 export/import API

ilammy opened the pull request:
This PR is work-in-progress on key export and import. Here we add low-level API for exporting and importing keys, and implement the export code path which is easier. Importing will come later as it's a bit harder. Meanwhile I'd like to get feedback on whatever is ready at the moment.

First of all we get rid of an unused key type for public keys only. I intended it to be used when a service needs to store only public keys of its peers. However, it's not convenient to use when programming, and it's not really a new "type" – it's just the same Themis key pair, just without the private part. So, instead of inventing a new type, just reuse the existing key pair and store only the public key. Update the code to allow skipping private keys and return a more appropriate error when a private key is requested for a key pair that does not have one.

The next commit adds the data structure used for exported keys. It's a simple list of key ring data which is going to be encrypted and signed for export. Resulting binary blob is what's going to be stored for backup and transferred between machines. Its structure is the following:

- `SignedContainer` signed with an exporting key (different from the key used by the key store)
  - binary blob, result of Themis Secure Cell encryption with another exporting key for encryption 
    - `EncryptedKeys` structure containing `KeyRing` data
      Individual key rings and keys inside them are not encrypted. Instead, the entire list is encrypted to conceal its contents.

Finally, we add some API that's going to be used for exporting and importing key rings. For export you give a list of key rings you need and get an encrypted blob in return. Then this blob can be used to import the key rings inside it. In both cases you need to explicitly pass a "crypto suite" which will be used for the transfer, it is different from the suite used by the key store to access the key data. 

Also, importing operation is more involved. We have to decide what to do if the key ring already exists. I currently have no concrete vision of what situations we might run into. For example, the same backup may be imported twice, or an outdated backup applied, or an incremental one, or there is an overlap in data. Whatever it is, I want the interface to be extensible enough to be able to handle that situations when we identify them.

This PR implements only the exporting path, and it's more like a backend support. `acra-keys` tool still has to be taught to use this API, and for that we'll need to first provide an API to retrieve all available key rings.

Lagovas commented on keystore/v2/keystore/filesystem/export.go:
```diff
@@ -0,0 +1,221 @@
+
+func (s *KeyStore) importKeyRing(newRingData *asn1.KeyRing, delegate *api.KeyRingImportDelegate) error {
+	keyRing := newKeyRing(s, string(newRingData.Purpose))
+	err := s.readKeyRing(keyRing)
```
can you add comment that here we check existing keyring with same purpose?) I wrote block of questions what we try to do until understood) because it's clear only when read block with `err == nil` below why we read empty keyring with same purpose)

ilammy replied:
> can you add comment that here we check existing keyring with same purpose?

Will do later. Right now `importKeyRing` is more of a draft, it will be expanded in the next PR. I hope it will be more readable then.

Lagovas commented on keystore/v2/keystore/api/keyStore.go:
```diff
@@ -34,4 +44,31 @@ type MutableKeyStore interface {
+)
+
+// KeyRingImportDelegate controls details of key ring import process.
+type KeyRingImportDelegate struct {
```
why to not use interface instead struct?

ilammy replied:
> why to not use interface instead struct?

Hm... The idea was that there could be more than one aspect of key import process that the user would like to control, so I wanted to make it possible to specify them independently, possibly without specifying some parts they don’t care about.

However, now the users typically have to construct a `KeyRingImportDelegate` and still fill it in with some predefined strategies. Your comment made me realize that we can really just use an interface for that and provide a bunch of premade strategies for the most common use cases. If that does not fit the bill then go ahead, implement the entire `KeyRingImportDelegate` interface the way you want it.

I’ll change this to an interface instead.

Lagovas commented on keystore/v2/keystore/filesystem/export.go:
```diff
@@ -0,0 +1,221 @@
+		// to the delegate and let it decide how to proceed.
+		decision := api.ImportAbort
+		err := ErrKeyRingExists
+		if delegate != nil && delegate.DecideKeyRingOverwrite != nil {
```
that is why struct seems like not a comfortable choice in use. imho better to provide interface for strategy pattern which will be used on importing when keyring already exists. And provide several types of strategy: "skip"/"overwrite"/"abort". And here just call `delegate.ImportExistingData` or something like that. Which delegate to use will be chosen on start by keystore configuration and some fabric method like:
```
switch importStrategy {
case api.ImportOverwrite:
  return NewImportOverwriteStrategy()
case api.ImportSkip:
  return NewImportSkipStrategy()
case api.AbortOnOverwrite:
  return NewAbortStrategy()
}
```

then here will be more clean code, one switch branch on start and zero in runtime execution here.

ilammy replied:
> And here just call delegate.ImportExistingData or something like that. 

That would look a bit neater from our side but that will couple the delegate implementation to the key store implementation, without a particularly good reason. If we use enumeration values then the delegate can be reused with different key store implementations without any issues.

Lagovas commented on keystore/v2/keystore/filesystem/export.go:
```diff
@@ -0,0 +1,221 @@
+		}
+	}()
+	for i := range exported.Keys {
+		decrypted, err := r.decryptAllKeyData(exported.Keys[i].Data, exported.Keys[i].Seqnum)
```
why we decrypt data for export? to encrypt it with "export" key and allow to encrypt with another key on import operation?

ilammy replied:
> why we decrypt data for export? to encrypt it with "export" key and allow to encrypt with another key on import operation?

Yes, that is correct. Suppose you’re transferring keys between key stores A and B. Each of them uses their own encryption keys and at the moment of export from A we don’t know where the keys will be imported to. During export we’ll decrypt them with key A, encrypt them back with a temporary key T. During import the T key will be transferred to the B place (in some unspecified, secure way), then the keys will be decrypted with T and encrypted for B with B’s key.

Lagovas commented on keystore/v2/keystore/filesystem/key.go:
```diff
@@ -261,23 +261,19 @@ func (r *KeyRing) addKeyData(data api.KeyData, key *asn1.Key) error {
-	case api.ThemisKeyPairFormat:
-		if len(data.PublicKey) == 0 || len(data.PrivateKey) == 0 {
-			return api.ErrNoKeyData
+		if len(data.PrivateKey) != 0 {
```
if `len(data.PrivateKey) == 0` we do nothing, but if empty public key, we return ErrNoKeyData? Why not same behavior for private key for `ThemisKePairFormat`?

ilammy replied:
> if `len(data.PrivateKey) == 0` we do nothing, but if empty public key, we return ErrNoKeyData? Why not same behavior for private key for ThemisKePairFormat?

Because it makes sense to have a public part of the key pair stored without a corresponding private key. We definitely don’t want to store neither of them. It also makes little sense to allow to store the private key alone, given how we use it with Themis. Themis allows to specify only private key for signature verification in Secure Message but we don’t really use it that way in Acra.

ilammy commented:
> @ilammy added 3 commits 19 days ago
> @Lagovas reviewed 11 days ago

Shamefur dispray 😡

ilammy commented on keystore/v2/keystore/api/tests/key.go:
```diff
@@ -190,13 +193,12 @@ func testKeyInvalidInputs(t *testing.T, newKeyStore NewKeyStore) {
 	})
 	if err != api.ErrNoKeyData {
-		t.Errorf("cannot add public key without data: %v", err)
+		t.Errorf("cannot add key pair without any data: %v", err)
```
> we can add key without any data for cases when key material removed but we store metadata for history, yes?

Not really. This test verifies that it is not allowed to add *only* key metadata, without any key material. The reasoning is that you should not be backdating metadata alone.

However, it is possible (or, well, should be possible) to add a key with metadata then at some point destroy the key material so that only metadata remains.

Though, it’s an open question how this should interact with key export/import. I don’t know whether we should allow exporting destroyed keys and importing them back (which will be metadata-only import).

ilammy commented on keystore/v2/keystore/api/tests/keyStore.go:
```diff
@@ -19,9 +19,110 @@ package tests
+	s := newKeyStore(t)
+	setupDemoKeyStore(s, t)
+
+	_, err := s.ExportKeyRings(exportRingAll, newExportStoreSuite(t))
```
> why not check that exported data is the same as expected?

It’s all encrypted so it’s hard to verify the contents. Right now this test is more of a placeholder to check that you can export stuff without an error. It will be properly verified in the next PR that adds import. Then we’d try importing the exported data and check that we get the key data we should get.

ilammy commented:
I've reviewed and updated the branch with the rest of key import implementation, and it turned out to be not that big. So I've just added it here instead of creating a new PR for the import specifically. @Lagovas, please give this PR another look.

The tests... well... they leave much to be desired in the conciseness department, but they do verify what they need to verify.

Lagovas commented on keystore/v2/keystore/filesystem/keyRingTX.go:
```diff
@@ -133,6 +133,34 @@ func (tx *txAddKey) Rollback(ring *KeyRing) error {
+}
+
+func (tx *txSetKeys) Rollback(ring *KeyRing) error {
+	if ring.data.Current != tx.oldCurrent {
```
is not it should be equal to `tx.current`? In `Apply` method we setup `current` for ring. In `Rollback` we should expect any other value from oldCurrent or just setup `oldCurrent`.
p.s. what means "OOO" in `errTxOOORollback`?)

vixentael replied:
> p.s. what means "OOO" in errTxOOORollback?)

cat on keyboard?

ilammy replied:
@Lagovas, I believe you’re right, that line is incorrect. If the current subkey does not change then we’ll get an error here.

> p.s. what means "OOO" in errTxOOORollback?)

Out-of-order.

> cat on keyboard?

I wish it were. The only cat I have is a ceramic one, it can definitely spam one keypress if it falls of the shelf onto the keyboard.

----

I’ve added these checks to verify that we do not roll back changes out-of-order, but this code path is not well tested (if tested at all). So I guess we can drop all these lines and always roll back to whatever values we think are correct ones?

Lagovas replied:
then better to rename to full name `errTxOutOfOrderRollback`. Because i'm not sure that anyone of us will be able to repeat what it means after 1-3 months)

and yes, rollback should return state of something to state when it was at start. In this case we store state at start of the transaction and know which state it should take. And it doesn't need to be changed always on rollback. For example with databases, state of table can be unchanged but the transaction rollbacked by some other condition and it's okay. There the same, we don't need check that something changed.
